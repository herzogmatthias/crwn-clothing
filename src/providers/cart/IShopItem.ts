export interface IShopItem {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
}
