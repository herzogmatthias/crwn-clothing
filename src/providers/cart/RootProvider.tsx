import * as React from "react";
import CartProvider from "./cart.provider";

export interface IRootProviderProps {
  children: React.ReactNode;
}

export default function RootProvider({ children }: IRootProviderProps) {
  return (
    <div>
      <CartProvider>{children}</CartProvider>
    </div>
  );
}
