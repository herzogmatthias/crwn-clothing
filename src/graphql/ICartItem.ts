import { IShopItem } from "./IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
