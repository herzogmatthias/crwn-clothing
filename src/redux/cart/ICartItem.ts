import { IShopItem } from "../shop/IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
