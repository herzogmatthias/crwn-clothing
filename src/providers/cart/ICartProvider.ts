import { ICartItem } from "../../redux/cart/ICartItem";
import { IShopItem } from "../../redux/shop/IShopItem";

export interface ICartProvider {
  hidden: boolean;
  total: number;
  toggleHidden(): void;
  cartItems: ICartItem[];
  addItem(item: IShopItem): void;
  removeItem(item: ICartItem): void;
  clearItemFromCart(item: ICartItem): void;
  cartItemsCount: number;
}
