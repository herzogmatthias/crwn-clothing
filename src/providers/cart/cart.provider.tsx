import React, { createContext, useEffect } from "react";
import { ICartProvider } from "./ICartProvider";
import { ICartItem } from "../../redux/cart/ICartItem";
import {
  addItemToCart,
  removeItemFromCart,
  filterItemFromCart,
  getCartItemsCount,
  getTotalPrice,
} from "../../redux/cart/cart.utils";
import { IShopItem } from "../../redux/shop/IShopItem";

export const CartContext = createContext<ICartProvider>({
  hidden: true,
  total: 0,
  cartItems: [],
  cartItemsCount: 0,
  toggleHidden: () => {},
  addItem: () => {},
  removeItem: () => {},
  clearItemFromCart: () => {},
});

interface ICartProviderProps {
  children: React.ReactNode;
}

const CartProvider = ({ children }: ICartProviderProps) => {
  const [hidden, setHidden] = React.useState(true);

  const [cartItems, setCartItems] = React.useState<ICartItem[]>([]);
  const [cartItemsCount, setCartItemsCount] = React.useState(0);

  const [total, setTotal] = React.useState(0);

  const toggleHidden = () => {
    setHidden(!hidden);
  };

  const addItem = (item: IShopItem) =>
    setCartItems(addItemToCart(cartItems, item));

  const removeItem = (item: ICartItem) =>
    setCartItems(removeItemFromCart(cartItems, item));
  const clearItemFromCart = (item: ICartItem) =>
    setCartItems(filterItemFromCart(cartItems, item));

  useEffect(() => {
    setCartItemsCount(getCartItemsCount(cartItems));
    setTotal(getTotalPrice(cartItems));
  }, [cartItems]);

  return (
    <CartContext.Provider
      value={{
        total,
        addItem,
        hidden,
        toggleHidden,
        cartItems,
        cartItemsCount,
        removeItem,
        clearItemFromCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export default CartProvider;
