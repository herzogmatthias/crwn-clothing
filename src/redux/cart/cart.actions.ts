import { createAction } from "@reduxjs/toolkit";
import { TOGGLE_CART_HIDDEN, ADD_ITEM } from "./cart.types";
import { withPayloadType } from "../genericActionPayloadType";
import { IShopItem } from "../../pages/shop/IShopItem";
import { ICartItem } from "./ICartItem";

export const toggleCartHidden = createAction(TOGGLE_CART_HIDDEN);
export const addItem = createAction(ADD_ITEM, withPayloadType<IShopItem>());
