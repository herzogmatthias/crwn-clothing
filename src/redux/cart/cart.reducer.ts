import { ICartState } from "./ICartState";
import { createReducer, PayloadAction } from "@reduxjs/toolkit";
import { toggleCartHidden, addItem } from "./cart.actions";
import { IShopItem } from "../../pages/shop/IShopItem";
import { addItemToCart } from "./cart.utils";

const initialState: ICartState = {
  hidden: true,
  cartItems: [],
};

const cartReducer = createReducer(initialState, {
  [toggleCartHidden.type]: (state, action) => {
    state.hidden = !state.hidden;
  },
  [addItem.type]: (state, action: PayloadAction<IShopItem>) => {
    state.cartItems = addItemToCart(state.cartItems, action.payload);
  },
});

export default cartReducer;
