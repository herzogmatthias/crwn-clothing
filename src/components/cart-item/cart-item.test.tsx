import React from "react";
import { shallow } from "enzyme";
import CartItem from "./cart-item.component";
import { ICartItem } from "../../redux/cart/ICartItem";

describe("NON CONNECTED CARTITEM TESTS", () => {
  let item: ICartItem = {
    id: 1,
    imageUrl: "asdf",
    name: "hats",
    price: 34,
    quantity: 2,
  };
  let wrapper = shallow(<CartItem item={item}></CartItem>);
  test("render CartItem Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
