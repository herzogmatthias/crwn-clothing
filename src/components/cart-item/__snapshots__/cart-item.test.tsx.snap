// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NON CONNECTED CARTITEM TESTS render CartItem Component 1`] = `
<styled.div>
  <styled.img
    alt="item"
    src="asdf"
  />
  <styled.div>
    <styled.span>
      hats
    </styled.span>
    <span
      className="price"
    >
      2
       x $
      34
    </span>
  </styled.div>
</styled.div>
`;
